### PR TITLE
Add cluster label to shard metric

### PR DESCRIPTION
**Pull request by arpitjindal97**
Almost all the other metrics have this label, so it becomes a norm to have it on this metric as well

### CODE REVIEW

**Comment by arpitjindal97** [comment]
@sysadmind Can you please have a look at this?

File: collector/shards.go [hunk 1]
```diff
@@ -156,20 +159,26 @@ func (s *Shards) Collect(ch chan<- prometheus.Metric) {
 	nodeShards := make(map[string]float64)
 
 	for _, shard := range sr {
-		if val, ok := nodeShards[shard.Node]; ok {
-			nodeShards[shard.Node] = val + 1
-		} else {
-			nodeShards[shard.Node] = 1
+		if shard.State == "STARTED" {
```
Comment (sysadmind) [inline collector/shards.go:162]: This seems to change the behavior of the metric. What is the intention here?

**Review by sysadmind — COMMENTED** [review]
There are a lot of metrics that don't include the cluster name. I'm not sure that it makes sense to create a `Nodes{}` object and pull that data again just to add a label. I think to accept this we would need a better way to share that data.

**Comment by arpitjindal97** [comment]
So, we have like 100 of ElasticSearch clusters and we are running this exporter in each node and collecting metrics in a single prometheus storage.

We want `cluster` label so that we can distinguish between multiple clusters. We have one grafana dashboard with `cluster` as a variable which we are passing in queries. This specific metric doesn't have it, so I raised the Pull Request.

File: collector/shards.go (hunk 1 above)
Comment (arpitjindal97) [reply collector/shards.go:162]: Ideally, this metrics should be able to tell how many shards each node is handling. The api which this exporter hit is `_cat/shards` which also lists down the relocating shards.

image

Those entries are disruptive and according to existing source code, `node` label gets that whole big string with an arrow and ip address which doesn't make sense.

So, It's better to only capture already relocated shards aka (STARTED)

**Comment by sysadmind** [comment]
I'm on board with this change, but I would like to remove the dependency on `Nodes` to get the cluster name. There is some existing work to have collectors receive updates to the cluster info: https://github.com/prometheus-community/elasticsearch_exporter/blob/master/main.go#L208. It's not the prettiest functionality, but it will cut down on API calls for those user that care about that type of thing. It would also mean if we refactor in the future, we have a common set of functionality to consider.

**Comment by arpitjindal97** [comment]
@sysadmind I have updated the code. Please review

**Comment by arpitjindal97** [comment]
@SuperQ  Can you please review it?

File: collector/shards.go [hunk 2]
```diff
@@ -38,52 +40,100 @@ var (
...
-	Labels func(node string) []string
+	Labels labels
 }
 
 // NewShards defines Shards Prometheus metrics
 func NewShards(logger log.Logger, client *http.Client, url *url.URL) *Shards {
-	return &Shards{
+
+	nodeLabels := labels{
+		keys: func(...string) []string {
```
Comment (sysadmind) [inline collector/shards.go:81]: Why is this a func if returns a static value? Why does it take parameters if it ignores them?

File: collector/shards.go [hunk 3]
```diff
@@ -38,52 +40,100 @@ var (
...
+			return []string{"node", "cluster"}
```
Comment (SuperQ) [inline collector/shards.go:82]: Can we name this label `elastic_cluster`? `cluster` is commonly used as an external label.

**Review by SuperQ — CHANGES_REQUESTED** [review]

File: main.go [hunk 4]
```diff
@@ -202,13 +202,18 @@ func main() {
 	prometheus.MustRegister(collector.NewNodes(logger, httpClient, esURL, *esAllNodes, *esNode))
 
 	if *esExportIndices || *esExportShards {
-		prometheus.MustRegister(collector.NewShards(logger, httpClient, esURL))
+		sC := collector.NewShards(logger, httpClient, esURL)
+		prometheus.MustRegister(sC)
 		iC := collector.NewIndices(logger, httpClient, esURL, *esExportShards, *esExportIndexAliases)
 		prometheus.MustRegister(iC)
-		if registerErr := clusterInfoRetriever.RegisterConsumer(iC); registerErr != nil {
+		if clusterInfoRetriever.RegisterConsumer(iC) != nil {
```
Comment (sysadmind) [inline main.go:209]: This isn't idiomatic go. 
```suggestion
		if registerErr := clusterInfoRetriever.RegisterConsumer(iC); registerErr != nil {
```

File: main.go [hunk 5]
```diff
@@ -202,13 +202,18 @@ func main() {
...
 			level.Error(logger).Log("msg", "failed to register indices collector in cluster info")
 			os.Exit(1)
 		}
+		if clusterInfoRetriever.RegisterConsumer(sC) != nil {
```
Comment (sysadmind) [inline main.go:213]: again, this isn't idiomatic
```suggestion
		if registerErr := clusterInfoRetriever.RegisterConsumer(sC); registerErr != nil {
```

**Review by sysadmind — CHANGES_REQUESTED** [review]

File: collector/shards.go (hunk 2 above)
Comment (arpitjindal97) [reply collector/shards.go:81]: It's done like this in indices.go

File: collector/shards.go (hunk 3 above)
Comment (arpitjindal97) [reply collector/shards.go:82]: All the other metrics have cluster label, If we wish to change it to `elastic_cluster` then it should be done in all metrics

**Comment by sysadmind** [comment]
Your comment/screenshot is for `_cat/shards`, but the exporter doesn't use this URL in that way. Your screenshot shows data in a formay that the exporter doesn't see. The URL that the exporter uses is `_cat/shards?format=json`

File: collector/shards.go (hunk 1 above)
Comment (arpitjindal97) [reply collector/shards.go:209]: In the JSON, I see it's showing `UNASSIGNED` shards also. We only should be capturing `STARTED` shards

image

Comment (arpitjindal97) [reply collector/shards.go:209]: if shards are RELOCATING. this is how it appears:
image

Comment (arpitjindal97) [reply collector/shards.go:209]: This metric only intents to tell how many shards each node currently have. Therefore, we must exclude `RELOCATING` & `UNASSIGNED`

**Comment by arpitjindal97** [comment]
@sysadmind I have added the screenshots from correct API.

File: collector/shards.go (hunk 3 above)
Comment (SuperQ) [reply collector/shards.go:82]: Yea, seems like a good idea to me. I didn't know the other metrics had the same problem. We can do that as a separate PR if you like.

**Comment by arpitjindal97** [comment]
@sysadmind @SuperQ Please review

**Review by SuperQ — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,4 @@ elasticsearch_exporter
 *-stamp
 .tarballs
 /vendor
+/.idea
```
[the remark this answers is not in the document]
Comment (arpitjindal97) [reply .gitignore]: Fixed

**Comment by arpitjindal97** [comment]
@sysadmind  @SuperQ  Please review and merge

**Comment by arpitjindal97** [comment]
@sysadmind Please review. I'm really blocked by this

**Comment by arpitjindal97** [comment]
@sysadmind @SuperQ Gentle reminder

**Review by sysadmind — APPROVED** [review]
LGTM. The remaining feedback is all around the refactoring of collectors that we are working on, so that can be addressed as part of that effort.

**Pull request merged**